src/vm/tests/principals.rs

```diff
@@ -178,3 +182,353 @@ fn test_simple_is_standard_undefined_cases() {
+        execute_against_version_and_network(input, ClarityVersion::Clarity2, false)
+    );
+
+    // u22 is not a byte buffer, so is invalid. Should have been caught by type checker so use
+    // `CheckErrors`.
+    let input = r#"(principal-construct u22 0x0102030405060708091011121314151617181920)"#;
+    assert_eq!(
+        Err(CheckErrors::TypeValueError(
+            SequenceType(BufferType(BufferLength(1))),
+            Value::UInt(22)
+        )
+        .into()),
+        execute_against_version_and_network(input, ClarityVersion::Clarity2, false)
+    );
+
+    // Hash key part is too large, should have length 20. This is a `CheckErrors` error because it
+    // should have been caught by the type checker.
+    let input = r#"(principal-construct 0x16 0x010203040506070809101112131415161718192021)"#;
+    assert_eq!(
+        execute_against_version_and_network(input, ClarityVersion::Clarity2, false).unwrap_err(),
+        CheckErrors::TypeValueError(
+            SequenceType(BufferType(BufferLength(20))),
+            Value::Sequence(SequenceData::Buffer(BuffData {
+                data: hex_bytes("010203040506070809101112131415161718192021").unwrap()
+            }))
+        )
+        .into()
+    );
+}
+
+#[test]
+// Test cases where we return an "in response" error.
+fn test_principal_construct_response_errors() {
+    // Hash key part is too small, should have length 20. This wasn't for the type checker, so the
+    // error is signaled in the returned Response.
+    let input = r#"(principal-construct 0x16 0x01020304050607080910111213141516171819)"#;
+    assert_eq!(
+        execute_against_version_and_network(input, ClarityVersion::Clarity2, false)
+            .unwrap()
+            .unwrap(),
+        Value::Response(ResponseData {
+            committed: false,
+            data: Box::new(Value::Tuple(
+                TupleData::from_data(vec![
+                    ("error_int".into(), Value::UInt(1 as u128)),
+                    ("value".into(), Value::none()),
+                ])
+                .expect("FAIL: Failed to initialize tuple."),
+            )),
+        }),
+    );
+
+    // Version byte is too small, should have length 1. This error is signaled in the returned
+    // Response.
+    let input = r#"(principal-construct 0x 0x0102030405060708091011121314151617181920)"#;
+    assert_eq!(
+        execute_against_version_and_network(input, ClarityVersion::Clarity2, false)
+            .unwrap()
+            .unwrap(),
+        Value::Response(ResponseData {
+            committed: false,
+            data: Box::new(Value::Tuple(
+                TupleData::from_data(vec![
+                    ("error_int".into(), Value::UInt(1 as u128)),
+                    ("value".into(), Value::none()),
+                ])
+                .expect("FAIL: Failed to initialize tuple."),
+            )),
+        }),
+    );
+
+    // The version byte 0x20 is too big, even for the future. So, we get no result.
```

Comment:
I would bet Bitcoins to Satoshis that 32 address bytes is not going to be enough.

Reply:
I thought you could only represent 32 options with the "c32" address.

Reply:
The version *byte* can encode up to 256 possible value.